test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -79,167 +87,320 @@ func (s *snapshottableTestSuite) SkipRedundantSuite(driver TestDriver, pattern t
 		framework.ExpectNoError(err)
 
 		// Get the bound PV
-		_, err = cs.CoreV1().PersistentVolumes().Get(context.TODO(), pvc.Spec.VolumeName, metav1.GetOptions{})
+		_, err = l.cs.CoreV1().PersistentVolumes().Get(context.TODO(), l.pvc.Spec.VolumeName, metav1.GetOptions{})
 		framework.ExpectNoError(err)
+	}
+	cleanup := func(l *snapshottableLocal) {
+		// All tests will require these driver cleanup tests
+		err := tryFunc(l.driverCleanup)
+		framework.ExpectNoError(err, "while cleaning up driver")
+
+		// Depending on how far the test executed, cleanup accordingly
+		// Execute in reverse order, similar to defer stack
+		for i := len(l.cleanupSteps) - 1; i >= 0; i-- {
+			l.cleanupSteps[i]()
+		}
+	}
 
-		ginkgo.By("creating a SnapshotClass")
-		vsc, err = dc.Resource(SnapshotClassGVR).Create(context.TODO(), vsc, metav1.CreateOptions{})
-		framework.ExpectNoError(err)
-		defer func() {
-			framework.Logf("deleting SnapshotClass %s", vsc.GetName())
-			framework.ExpectNoError(dc.Resource(SnapshotClassGVR).Delete(context.TODO(), vsc.GetName(), metav1.DeleteOptions{}))
-		}()
+	ginkgo.BeforeEach(func() {
+		// Check preconditions.
+		framework.ExpectEqual(pattern.SnapshotType, testpatterns.DynamicCreatedSnapshot)
+		dInfo := driver.GetDriverInfo()
+		ok := false
+		sDriver, ok = driver.(SnapshottableTestDriver)
+		if !dInfo.Capabilities[CapSnapshotDataSource] || !ok {
+			e2eskipper.Skipf("Driver %q does not support snapshots - skipping", dInfo.Name)
+		}
+		dDriver, ok = driver.(DynamicPVTestDriver)
+		if !ok {
+			e2eskipper.Skipf("Driver %q does not support dynamic provisioning - skipping", driver.GetDriverInfo().Name)
+		}
+	})
 
-		ginkgo.By("creating a snapshot")
-		snapshot := getSnapshot(pvc.Name, pvc.Namespace, vsc.GetName())
+	// This intentionally comes after checking the preconditions because it
+	// registers its own BeforeEach which creates the namespace. Beware that it
+	// also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
+	ginkgo.It("should create snapshot with delete policy [Feature:VolumeSnapshotDataSource]", func() {
+		l := &snapshottableLocal{}
 
-		snapshot, err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Create(context.TODO(), snapshot, metav1.CreateOptions{})
-		framework.ExpectNoError(err)
-		defer func() {
-			framework.Logf("deleting snapshot %q/%q", snapshot.GetNamespace(), snapshot.GetName())
-			// typically this snapshot has already been deleted
-			err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Delete(context.TODO(), snapshot.GetName(), metav1.DeleteOptions{})
-			if err != nil && !apierrors.IsNotFound(err) {
-				framework.Failf("Error deleting snapshot %q. Error: %v", pvc.Name, err)
-			}
-		}()
-		err = WaitForSnapshotReady(dc, snapshot.GetNamespace(), snapshot.GetName(), framework.Poll, framework.SnapshotCreateTimeout)
-		framework.ExpectNoError(err)
+		init(l)
+		defer cleanup(l)
```

Comment:
What if we hit an error (for e.g framework.ExpectNoError()) in one of the steps of the the init() call itself, in that case, we will miss cleanup of resources?
say for e.g. in init() we hit error after pod created or pv created. Can you please try it out?

Reply:
I'm not sure what we can do in that case. This will definitely happen in that case. I wanted to continue the execution of the cleanup steps, in that case, to continue getting rid of other resources.

Reply:
There were some changes done to other storage test suites so that in the cleanup, we gather all the errors, and then at the end finally throw an error. That way we make sure we have run through all cleanup steps. [Example](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/provisioning.go#L176)

Reply:
Using tryFunc inside of  `cleanup()` now.